File: code/python/topKFrequentWords.py
class Solution:
    def topKFrequentWords(self, words: List[str], k: int) -> List[str]:
        words.sort()
        dic1 = {}
        dic2 = {}
        i = 0
        Max = 0
        while i < len(words):
            if words[i] not in dic1:
                dic1[words[i]] = 1
            else:
                dic1[words[i]] += 1
            i = i + 1
          
        for key in dic1:
            if dic1[key] not in dic2:
                dic2[dic1[key]] = [key]
            else:
                dic2[dic1[key]].append(key) 
            if Max < dic1[key]:
                Max = dic1[key]
        res = []
        while Max > 0:
            if Max in dic2:
                res.extend(dic2[Max])
            if len(res) >=k:
                return res[:k]
            Max -= 1
        
        return res[:k]
                

File: code/java/topKFrequentWords.java
class Solution {
    public List<String> topKFrequentWords(String[] words, int k) {
        //Arrays.sort(words);
        Map<String,Integer> map1 = new HashMap<String,Integer>();
        List<String>[] bucket = new List[words.length+1];
        List<String> res = new ArrayList<String>();
        List<String> temp = new ArrayList<String>();
        for(String word:words)
            map1.put(word,map1.getOrDefault(word,0)+1);
        
        for(String key : map1.keySet()){
            int fre = map1.get(key);
            if(bucket[fre] == null)
                bucket[fre] = new ArrayList<String>();
            bucket[fre].add(key);
        }
        int i = words.length;
        int count = 0;
        while(i > 0){
            if(bucket[i] != null){
                temp = bucket[i];
                Collections.sort(temp);
                res.addAll(temp);
            }
            if(res.size() >= k)
                break;
            i--;
        }
        while(res.size() > k){
            int length = res.size();
            res.remove(length-1);
        }
            
        return res;        
    }
}
